README.md
# gr-homework

A program that stores, sorts, and displays simple records of people, with queryable REST endpoints.

By Chris Collazo

## Installation

Simply extract or clone gr-homework to its own directory, open a terminal and "cd gr-homework", and follow the usage below.

## Usage

This program reads in records of people, containing five specific fields and formatted via one of the following three separator schemes:

	LastName|FirstName|Gender|FavoriteColor|BirthDate
	LastName,FirstName,Gender,FavoriteColor,BirthDate
	LastName FirstName Gender FavoriteColor BirthDate

Each field must be a single continuous string. Gender is expected to be "male" or "female", though it is not case sensitive. Birthdates must be ENTERED in "mm-dd-yyyy" format, though the program eventually DISPLAYS them in "m/d/yyyy" format.

Some examples:

	Earhart|Amelia|female|blue|07-24-1897
	Washington,George,male,red,02-22-1732
	Lovelace Ada female purple 12-10-1815
	
There are two methods for using this program. The first is a command line usage which grabs files of records. It can be used like so, with the example files in "res/" (records MUST be separated by newlines):

    $ lein run [files]
    $ lein run "res/commas.txt"
    $ lein run "res/commas.txt" "res/spaces.txt"
    $ lein run "res/commas.txt" "res/spaces.txt" "res/pipes.txt"

This will scoop up all the records contained within and display them sorted in three orders: By gender (female first), then last name ascending; by birthdate, ascending; and by last name, descending.

The second way to use this program is via REST endpoints. First, start the server:

	$ lein run

This starts the server listening on port 8080. Then, use a program such as curl to send record lines to the POST endpoint at /records:

	$ curl -d 'Earhart|Amelia|female|blue|07-24-1897' 127.0.0.1:8080/records
	$ curl -d 'Washington,George,male,red,02-22-1732' 127.0.0.1:8080/records
	$ curl -d 'Lovelace Ada female purple 12-10-1815' 127.0.0.1:8080/records

NOTE: For this curl method to work, the records MUST be surrounded by SINGLE quotes!

You may now query the GET endpoints to get sorted record collections, in pretty-printed JSON format:

	$ curl 127.0.0.1:8080/gender
	$ curl 127.0.0.1:8080/name
	$ curl 127.0.0.1:8080/birthdate

You may optionally store these responses in files for later viewing:

	$ curl 127.0.0.1:8080/gender > gender.json
	$ curl 127.0.0.1:8080/name > name.json
	$ curl 127.0.0.1:8080/birthdate > birthdate.json

To end the server, in the terminal where it is running, simply enter Ctrl-C.

To run the test suite, simply use

	$ lein test

## License

Copyright © 2017 Chris Collazo


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-03-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-03-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/gr-homework/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gr-homework/compare/0.1.0...0.1.1


test/gr_homework/core_test.clj
;; Copyright © 2017 Chris Collazo

(ns gr-homework.core-test
  (:use gr-homework.core
        gr-homework.sorting)
  (:require [clojure.test :refer :all]
            [clj-time.core :as t]
            [ring.util.response :refer [response not-found]]
            [cheshire.core :as json]))

;; A sample set of records, these are also used in other test namespaces
(def example-records
  [{:last "Collazo" 
    :first "Christopher"
    :gender "male"
    :fav-color "green"
    :dob (t/local-date 1991 6 21)}
   {:last "Washington"
    :first "George"
    :gender "male"
    :fav-color "red"
    :dob (t/local-date 1732 2 22)}
   {:last "Earhart"
    :first "Amelia"
    :gender "female"
    :fav-color "blue"
    :dob (t/local-date 1897 7 24)}
   {:last "Ritchie"
    :first "Dennis"
    :gender "male"
    :fav-color "yellow"
    :dob (t/local-date 1941 9 9)}
   {:last "Lovelace"
    :first "Ada"
    :gender "female"
    :fav-color "purple"
    :dob (t/local-date 1815 12 10)}
   {:last "Randall"
    :first "Lisa"
    :gender "female"
    :fav-color "orange"
    :dob (t/local-date 1962 6 18)}])

(deftest records-response-test
  (testing "Ensuring our GET responses are consistent with the state
            of the records atom."
    (is (= (records-response sort-by-dob-ascending)
           (response "No records stored.\n")))
    (reset! records example-records)
    (is (= (:body (records-response sort-by-dob-ascending))
           (slurp "res/birthdate.json")))
    (is (= (:body (records-response sort-by-last-name-ascending))
           (slurp "res/name.json")))
    (is (= (:body (records-response sort-by-gender-then-last-name-ascending))
           (slurp "res/gender.json")))))

(deftest records-response-test
  (testing "Ensuring our POST action responses are consistent."
    (is (= (:body (post-record "Ritchie|Dennis|male|yellow|09-09-1941"))
           "Added record for Dennis Ritchie.\n"))
    (is (= (:body (post-record "Randall Lisa female orange 06-18-1962"))
           "Added record for Lisa Randall.\n"))
    (is (= (:body (post-record "Washington,George,male,red,02-22-1732"))
           "Added record for George Washington.\n"))
    (is (= (:status (post-record "Randall;Lisa;female;orange;06-18-1962"))
           500))
    (is (= (:status (post-record "RandallLisafemaleorange06-18-1962"))
           500))
    (is (= (:status (post-record "Ran Dall Lisa female orange 06-18-1962"))
           500))
    (is (= (:status (post-record "Ran Dall,Lisa,female,orange,06-18-1962"))
           500))))


test/gr_homework/parsing_test.clj
;; Copyright © 2017 Chris Collazo

(ns gr-homework.parsing-test
  (:require [clojure.test :refer :all]
            [gr-homework.parsing :refer :all]
            [clj-time.core :as t]))

(deftest determine-separator-test
  (testing "Testing that we can correctly determine separators."
    (is (= (str (determine-separator "Ritchie|Dennis|male|yellow|09-09-1941"))
           (str #"\|")))
    (is (= (str (determine-separator "Rit Chie|Dennis|male|yellow|09-09-1941"))
           (str #" ")))
    (is (= (str (determine-separator "Washington,George,male,red,02-22-1732"))
           (str #",")))
    (is (= (str (determine-separator "Randall Lisa female orange 06-18-1962"))
           (str #" ")))
    (is (thrown?
         Exception
         (determine-separator "Randall;Lisa;female;orange;06-18-1962")))
    (is (thrown?
         Exception
         (determine-separator "RandallLisafemaleorange06-18-1962")))))

(deftest parse-date-test
  (testing "Testing that we can correctly parse dates."
    (is (= (parse-date "09-09-1941")
           (t/local-date 1941 9 9)))
    (is (= (parse-date "02-22-1732")
           (t/local-date 1732 2 22)))
    (is (= (parse-date "06-18-1962")
           (t/local-date 1962 6 18)))
    (is (thrown?
         Exception
         (parse-date "09/09/1941")))
    (is (thrown?
         Exception
         (parse-date "09-1941-09")))
    (is (thrown?
         Exception
         (parse-date "1941-09-09")))))

(deftest parse-record-test
  (testing "Testing that we can correctly parse records."
    (is (= (parse-record "Ritchie|Dennis|male|yellow|09-09-1941")
           {:last "Ritchie"
            :first "Dennis"
            :gender "male"
            :fav-color "yellow"
            :dob (t/local-date 1941 9 9)}))
    (is (= (parse-record "Randall Lisa female orange 06-18-1962")
           {:last "Randall"
            :first "Lisa"
            :gender "female"
            :fav-color "orange"
            :dob (t/local-date 1962 6 18)}))
    (is (= (parse-record "Washington,George,male,red,02-22-1732")
           {:last "Washington"
            :first "George"
            :gender "male"
            :fav-color "red"
            :dob (t/local-date 1732 2 22)}))
    (is (thrown?
         Exception
         (parse-record "Randall;Lisa;female;orange;06-18-1962")))
    (is (thrown?
         Exception
         (parse-record "RandallLisafemaleorange06-18-1962")))
    (is (thrown?
         Exception
         (parse-record "Ran Dall Lisa female orange 06-18-1962")))
    (is (thrown?
         Exception
         (parse-record "Ran Dall,Lisa,female,orange,06-18-1962")))))

(deftest parse-records-file-test
  (testing "Testing that we can correctly parse a file."
    (is (= (parse-records-file "res/commas.txt")
           [{:last "Collazo"
             :first "Christopher"
             :gender "male"
             :fav-color "green"
             :dob (t/local-date 1991 6 21)}
            {:last "Washington"
             :first "George"
             :gender "male"
             :fav-color "red"
             :dob (t/local-date 1732 2 22)}]))
    (is (= (parse-records-file "res/pipes.txt")
           [{:last "Earhart"
             :first "Amelia"
             :gender "female"
             :fav-color "blue"
             :dob (t/local-date 1897 7 24)}
            {:last "Ritchie"
             :first "Dennis"
             :gender "male"
             :fav-color "yellow"
             :dob (t/local-date 1941 9 9)}]))
    (is (= (parse-records-file "res/spaces.txt")
           [{:last "Lovelace"
             :first "Ada"
             :gender "female"
             :fav-color "purple"
             :dob (t/local-date 1815 12 10)}
            {:last "Randall"
             :first "Lisa"
             :gender "female"
             :fav-color "orange"
             :dob (t/local-date 1962 6 18)}]))
    (is (thrown?
         Exception
         (parse-records-file "res/semicolons.txt")))))

(deftest parse-multiple-records-files-test
  (testing "Testing that we can correctly parse multiple files."
    (is (= (parse-multiple-records-files ["res/commas.txt"
                                          "res/pipes.txt"
                                          "res/spaces.txt"])
           [{:last "Collazo"
             :first "Christopher"
             :gender "male"
             :fav-color "green"
             :dob (t/local-date 1991 6 21)}
            {:last "Washington"
             :first "George"
             :gender "male"
             :fav-color "red"
             :dob (t/local-date 1732 2 22)}
            {:last "Earhart"
             :first "Amelia"
             :gender "female"
             :fav-color "blue"
             :dob (t/local-date 1897 7 24)}
            {:last "Ritchie"
             :first "Dennis"
             :gender "male"
             :fav-color "yellow"
             :dob (t/local-date 1941 9 9)}
            {:last "Lovelace"
             :first "Ada"
             :gender "female"
             :fav-color "purple"
             :dob (t/local-date 1815 12 10)}
            {:last "Randall"
             :first "Lisa"
             :gender "female"
             :fav-color "orange"
             :dob (t/local-date 1962 6 18)}]))
    (is (thrown?
         Exception
         (parse-multiple-records-files ["res/semicolons.txt"
                                        "res/slashes.txt"
                                        "res/pipes.txt"])))))


test/gr_homework/sorting_test.clj
;; Copyright © 2017 Chris Collazo

(ns gr-homework.sorting-test
  (:require [clojure.test :refer :all]
            [gr-homework.sorting :refer :all]
            [gr-homework.core-test :refer [example-records]]
            [clj-time.core :as t])) 

(deftest female?-test
  (testing "Testing that we correctly determine that a record is female."
    (is (= (female? (example-records 0)) false))
    (is (= (female? (example-records 2)) true))
    (is (= (female? nil) false))))

(deftest male?-test
  (testing "Testing that we correctly determine that a record is male."
    (is (= (male? (example-records 0)) true))
    (is (= (male? (example-records 2)) false))
    (is (= (male? nil) true))))

(deftest sort-by-last-name-ascending-test
  (testing "Testing that we correctly sort by last name, ascending."
    (is (= (sort-by-last-name-ascending example-records)
           [{:last "Collazo"  
             :first "Christopher"
             :gender "male"
             :fav-color "green"
             :dob (t/local-date 1991 6 21)}
            {:last "Earhart"
             :first "Amelia"
             :gender "female"
             :fav-color "blue"
             :dob (t/local-date 1897 7 24)}
            {:last "Lovelace"
             :first "Ada"
             :gender "female"
             :fav-color "purple"
             :dob (t/local-date 1815 12 10)}
            {:last "Randall"
             :first "Lisa"
             :gender "female"
             :fav-color "orange"
             :dob (t/local-date 1962 6 18)}
            {:last "Ritchie"
             :first "Dennis"
             :gender "male"
             :fav-color "yellow"
             :dob (t/local-date 1941 9 9)}
            {:last "Washington"
             :first "George"
             :gender "male"
             :fav-color "red"
             :dob (t/local-date 1732 2 22)}]))
    (is (= (sort-by-last-name-ascending nil) '()))))

(deftest sort-by-last-name-descending-test
  (testing "Testing that we correctly sort by last name, descending."
    (is (= (sort-by-last-name-descending example-records)
           [{:last "Washington"
             :first "George"
             :gender "male"
             :fav-color "red"
             :dob (t/local-date 1732 2 22)}
            {:last "Ritchie"
             :first "Dennis"
             :gender "male"
             :fav-color "yellow"
             :dob (t/local-date 1941 9 9)}
            {:last "Randall"
             :first "Lisa"
             :gender "female"
             :fav-color "orange"
             :dob (t/local-date 1962 6 18)}
            {:last "Lovelace"
             :first "Ada"
             :gender "female"
             :fav-color "purple"
             :dob (t/local-date 1815 12 10)}
            {:last "Earhart"
             :first "Amelia"
             :gender "female"
             :fav-color "blue"
             :dob (t/local-date 1897 7 24)}
            {:last "Collazo"  
             :first "Christopher"
             :gender "male"
             :fav-color "green"
             :dob (t/local-date 1991 6 21)}]))
    (is (= (sort-by-last-name-descending nil) '()))))

(deftest sort-by-gender-then-last-name-ascending-test
  (testing "Testing that we correctly sort by gender, then
            last name ascending."
    (is (= (sort-by-gender-then-last-name-ascending example-records)
           [{:last "Earhart"
             :first "Amelia"
             :gender "female"
             :fav-color "blue"
             :dob (t/local-date 1897 7 24)}
            {:last "Lovelace"
             :first "Ada"
             :gender "female"
             :fav-color "purple"
             :dob (t/local-date 1815 12 10)}
            {:last "Randall"
             :first "Lisa"
             :gender "female"
             :fav-color "orange"
             :dob (t/local-date 1962 6 18)}
            {:last "Collazo"  
             :first "Christopher"
             :gender "male"
             :fav-color "green"
             :dob (t/local-date 1991 6 21)}
            {:last "Ritchie"
             :first "Dennis"
             :gender "male"
             :fav-color "yellow"
             :dob (t/local-date 1941 9 9)}
            {:last "Washington"
             :first "George"
             :gender "male"
             :fav-color "red"
             :dob (t/local-date 1732 2 22)}]))
    (is (= (sort-by-gender-then-last-name-ascending nil) '()))))

(deftest sort-by-dob-ascending-test
  (testing "Testing that we correctly sort by birth date, ascending."
    (is (= (sort-by-dob-ascending example-records)
           [{:last "Washington"
             :first "George"
             :gender "male"
             :fav-color "red"
             :dob (t/local-date 1732 2 22)}
            {:last "Lovelace"
             :first "Ada"
             :gender "female"
             :fav-color "purple"
             :dob (t/local-date 1815 12 10)}
            {:last "Earhart"
             :first "Amelia"
             :gender "female"
             :fav-color "blue"
             :dob (t/local-date 1897 7 24)}
            {:last "Ritchie"
             :first "Dennis"
             :gender "male"
             :fav-color "yellow"
             :dob (t/local-date 1941 9 9)}
            {:last "Randall"
             :first "Lisa"
             :gender "female"
             :fav-color "orange"
             :dob (t/local-date 1962 6 18)}
            {:last "Collazo"  
             :first "Christopher"
             :gender "male"
             :fav-color "green"
             :dob (t/local-date 1991 6 21)}]))
    (is (= (sort-by-dob-ascending nil) '()))))


src/gr_homework/parsing.clj
;; Copyright © 2017 Chris Collazo

(ns gr-homework.parsing
  (:require [clojure.string :as str]
            [clj-time.core :as t]))

(defn determine-separator
  "Returns the separator of a record as a pattern.
   Throws an exception if a pipe, comma, or space is not found."
  [line]
  (loop [line-left line]
    (condp = (first line-left)
      \| #"\|" 
      \, #"," 
      \space #" "
      nil (throw (Exception. (str "in determine-separator: No or "
                                  "unrecognized separators in record.")))
      (recur (rest line-left)))))

(defn parse-date
  "Parses a date string in MM-DD-YYYY into a clj-time
   LocalDate object. Throws if it doesn't find three elements."
  [date-str]
  (let [date-nums (mapv #(. Integer parseInt %) (str/split date-str #"-"))]
    (if (not= 3 (count date-nums))
      (throw (Exception. "in parse-date: Dates must be in MM-DD-YYYY format."))
      (t/local-date (date-nums 2) (date-nums 0) (date-nums 1)))))

(defn parse-record
  "Parses an individual record, or one line.
   Expects, in order, last name, first name, gender,
   favorite color, and date of birth."
  [line]
  (let [entries (str/split line (determine-separator line))]
    (if (not= (count entries) 5)
      (throw (Exception. (str "in parse-record: Invalid number "
                              "of entries in record, or bad separator.")))
      {:last (entries 0)
       :first (entries 1)
       :gender (str/lower-case (entries 2))
       :fav-color (entries 3)
       :dob (parse-date (entries 4))})))

(defn parse-records-file
  "Parses records in a file, must contain line separations.
   Returns a vector of records."
  [filename]
  (->> filename
       (slurp)
       (str/split-lines)
       (mapv parse-record)))

(defn parse-multiple-records-files
  "Parses multiple records files, concats them.
   Returns a vector of records."
  [filenames]
  (->> filenames
       (map parse-records-file)
       (reduce concat)
       (vec)))


src/gr_homework/printing.clj
;; Copyright © 2017 Chris Collazo

(ns gr-homework.printing
  (:use gr-homework.sorting)
  (:require [clj-time.core :as t]))

(defn date-str
  "Return a printable string from a clj-time date, in the m/d/yyyy format."
  [localdate]
  (str (t/month localdate) "/" (t/day localdate) "/" (t/year localdate)))

(defn record-str
  "Format a neat multiline representation of a record for standard output."
  [rec]
  (str "\nFirst name:      " (:first rec)
       "\nLast name:       " (:last rec)
       "\nGender:          " (:gender rec)
       "\nFavorite color:  " (:fav-color rec)
       "\nDate of birth:   " (date-str (:dob rec))
       "\n"))

(defn multiple-records-str
  "Concat any number of records for neat standard output."
  [records]
  (reduce str (map record-str records)))

(defn output1
  [records]
  (str "\nSorted by gender (female first), last name ascending:\n"
       (multiple-records-str
        (sort-by-gender-then-last-name-ascending records))))

(defn output2
  [records]
  (str "\nSorted by birth date, ascending:\n"
       (multiple-records-str
        (sort-by-dob-ascending records))))

(defn output3
  [records]
  (str "\nSorted by last name, descending:\n"
       (multiple-records-str
        (sort-by-last-name-descending records))))

(defn three-outputs-str
  "Prints the three specified sorted variations to standard output."
  [records]
  (str (output1 records)
       (output2 records)
       (output3 records)))


src/gr_homework/core.clj
;; Copyright © 2017 Chris Collazo

(ns gr-homework.core
  (:use [clojure.tools.namespace.repl :only (refresh)]
        gr-homework.parsing
        gr-homework.printing
        gr-homework.sorting)
  (:require [org.httpkit.server :as http]
            [compojure.core :refer [defroutes context GET POST ANY]]
            [ring.util.response :refer [response not-found]]
            [ring.middleware.content-type :refer [wrap-content-type]]
            [ring.middleware.reload :refer [wrap-reload]]
            [cheshire.core :as json])
  (:gen-class))

;; An atom to act as a live store for records
(def records (atom nil))

;; A general handler for our three GET actions
;; In a larger project, this would likely be organized in a GET namespace
(defn records-response
  "If there are no records stored, responds with a message indicating such.
   Otherwise, runs the passed-in sorting func on stored records,
   JSON encodes, sends the resulting JSON back to client."
  [sorting-func]
  (response
   (if (nil? @records)
     "No records stored.\n"
     (json/generate-string
      (->> (sorting-func @records)
           (map #(assoc % :dob (date-str (:dob %)))))
      {:pretty true}))))

;; Our single POST action's handler
;; In a larger project, this would likely be organized in a POST namespace
(defn post-record
  "Given a string of a raw record line, post it to the atom
   and return a response depending on success or failure."
  [line]
  (try
    (let [rec (parse-record line)]
      (swap! records conj rec)
      (response (str "Added record for "
                     (:first rec) " " (:last rec)
                     ".\n")))
    (catch Exception e
      {:status 500
       :body (str "Could not add record: Exception "
                  (.getMessage e)
                  "\n")})))

;;
;; HTTP-Kit/Ring/Compojure Setup and Config
;;

(defroutes gr-homework-routes
  (context "/records" req
           (POST "/" {body :body}
                 (post-record (slurp body)))
           (GET "/gender" req
                (records-response sort-by-gender-then-last-name-ascending))
           (GET "/birthdate" req
                (records-response sort-by-dob-ascending))
           (GET "/name" req
                (records-response sort-by-last-name-ascending)))
  (ANY "*" req (not-found "404 Not Found:\nCheck README.md for instructions!\n")))

(def app
  (-> gr-homework-routes
      (wrap-content-type)
      (wrap-reload)))

(defonce server (atom nil))

(defn stop-server []
  (when-not (nil? @server)
    (@server :timeout 100)
    (reset! server nil)))

(defn start-server []
  (when (nil? @server)
    (reset! server (http/run-server #'app {:port 8080}))))

(defn restart-server []
  (when-not (nil? @server)
    (do (stop-server) (start-server))))

(defn -main
  "If given nothing, starts the REST server.
   Otherwise, any number of files can be named. Their records
   will be printed to standard output sorted in the three specified sort
   configurations."
  ([& filenames]
   (println "\nReading" (count filenames) "input files...")
   (print
    (try
      (three-outputs-str (parse-multiple-records-files filenames))
      (catch Exception e
        (str "Couldn't parse a file: " (.getMessage e))))))
  ([]
   (start-server)
   (println "Server started.")))
